plugins/modules/win_domain_ou.ps1

```diff
@@ -0,0 +1,257 @@
+if ($null -eq $path){
+    if ($($all_ous | Measure-Object | Select-Object -ExpandProperty Count) -eq 1){
+        $matched = $all_ous.DistinguishedName -match "DC=.+"
+    }elseif ($($all_ous | Measure-Object | Select-Object -ExpandProperty Count) -gt 1) {
+        $matched = $all_ous[0].DistinguishedName -match "DC=.+"
+    }else{
+        $module.FailJson("Path was null and unable to determine default domain $($_.Exception.Message)", $_)
+    }
+    if ($matched){
+        $path = $matches.Values[0]
+    }else{
+        $module.FailJson("Unable to find default domain $($_.Exception.Message)", $_)
+    }
+}
+$module.Result.path = $path
+
+# determine if requested OU exist
+Try {
+    $current_ou = $all_ous | Where-Object {
+            $_.DistinguishedName -eq "OU=$name,$path"}
+    $module.Diff.before = $current_ou
+    $module.Result.ou = Get-OuObject $module.Diff.before
+} Catch {
+    $module.Diff.before = $null
+    $current_ou = $null
+}
+if ($state -eq "present") {
+    # ou does not exist, create object
+    if(-not $current_ou) {
+        $params.Name = $name
+        $params.Path = $path
+        $module.Result.params = $params
+        Try {
+            New-ADOrganizationalUnit @params @onboard_extra_args -ProtectedFromAccidentalDeletion $protected -WhatIf:$check_mode
+        }Catch {
+            $module.FailJson("Failed to create organizational unit: $($_.Exception.Message)", $_)
+        }
+    }
+    # ou exists, update object
+    if ($current_ou) {
+        Try {
+            Set-ADOrganizationalUnit -Identity "OU=$name,$path" @params @onboard_extra_args -WhatIf:$check_mode
+        }Catch {
+            $module.FailJson("Failed to update organizational unit: $($_.Exception.Message)", $_)
+        }
+    }
+}
+
+if ($state -eq "absent") {
+    # ou exists, delete object
+    if ($current_ou -and -not $check_mode) {
+        Try {
+            # override protected from accidental deletion
+            Set-ADOrganizationalUnit -Identity "OU=$name,$path" -ProtectedFromAccidentalDeletion $false @onboard_extra_args -Confirm:$False -WhatIf:$check_mode
+        }Catch{
+            $module.FailJson("Failed to remove ProtectedFromAccidentalDeletion Lock: $($_.Exception.Message)", $_)
+        }
+            # check recursive deletion
+        if ($recursive) {
+            try{
+                Remove-ADOrganizationalUnit -Identity "OU=$name,$path" -Confirm:$False -WhatIf:$check_mode -Recursive @onboard_extra_args
+            }catch{
+                $module.FailJson("Failed to recurssively Remove-ADOrganizationalUnit $($_.Exception.Message)", $_)
+            }
+        }else{
+            try{
+                Remove-ADOrganizationalUnit -Identity "OU=$name,$path" -Confirm:$False -WhatIf:$check_mode @onboard_extra_args
+            }Catch{
+                $module.FailJson("Failed to Remove-ADOrganizationalUnit: $($_.Exception.Message)", $_)
+            }
+        }
+    }
+}
+
+# determine if a change was made
```

Comment:
This shouldn't be needed at all, `$module.Result.changed = $true` should be set after every operation where a change will occur (where you call `New/Set/Remove-ADOrganizationalUnit`. The same applies with `$module.Diff.after` you should be adding that in the code where the changes are being applied rather than at the end here.

The biggest benefits is now you don't need to split the change detection logic for when you are in check mode and not and the diff could potentially be simplified in some cases (absent will just always set after as `""`.

Reply:
updated code to reflect what you said.  it did simplify much of the code.